yarnpkg/yarn
# Fix check of scoped packages' sub-dependencies

**Summary**

When running `yarn check` on a freshly installed project with an extensive use of scoped packages yarn complains that some scoped packages uses the wrong version of another scoped package.

***What is the current behavior?***

With a module structure like following:

```
node_modules/
  @scope/
    subdependency@2.x/
    majordependency/
      node_modules/
        @scope/
          subdependency@1.x/
```

The following complaint is received when running `yarn check` (confirmed with latest `master`):

```
error @scope/majordependency##@scope/subdependency@^1.0.0 doesn't satisfy found match of @scope#subdependency@2.0.0
```

***What is the expected behavior?***

With the above mentioned module structure it would be expected that `yarn check` instead passes, as `@scope/majordependency` has the correct version of `@scope/subdependency` installed beneath it.

***Cause of bug***

`yarn check` when splitting the paths of a dependency expects `node_modules` to be inbetween every path part, but for scoped packages there should be no `node_modules` folder inbetween the scope and package name parts.

So when `yarn check` looks for `node_modules/@scope/node_modules/majordependency/node_modules/@scope/subdependency/package.json` it should instead be looking for that very same path but with `@scope/majordependency` instead of `@scope/node_modules/majordependency` in it.

**Test plan**

1. In a project, add a scoped module, lets call it `@scope/subdependency`
2. Run `yarn check`. Everything should be fine both with and without this PR.
3. Add another scoped dependency that also relies on the module from `1.`, but which requires an incompatible version of it. Lets call this `@scope/majordependency`. You should now have a similar module setup as outlined above.
4. Run `yarn check`. It should now complain without this PR, but pass with this PR.
5. (Manually remove `node_modules/@scope/majordependency/node_modules/@scope/subdependency/package.json` and run `yarn check` again. It should now correctly complain about the missing dependency.)

**Additional notes**

I'm seeing an unrelated failed test in the current master locally, both with and without this PR, but I've ignored it since it shouldn't be related to this.

I've not added any new tests because I couldn't really find the place where the tests for `yarn check` are. If someone could point me in the right direction then I could happily look at adding some tests for this.
